kernel/src/scan/mod.rs

```diff
@@ -440,6 +447,111 @@ impl Scan {
+                StructField::nullable("cardinality", DataType::LONG),
+            ]);
+            DataType::struct_type(vec![StructField::nullable(
+                "add",
+                DataType::struct_type(vec![
+                    StructField::nullable("path", DataType::STRING),
+                    StructField::nullable("partitionValues", partition_values),
+                    StructField::nullable("size", DataType::LONG),
+                    StructField::nullable("modificationTime", DataType::LONG),
+                    StructField::nullable("stats", DataType::STRING),
+                    StructField::nullable("deletionVector", deletion_vector),
+                ]),
+            )])
+        });
+
+        if hint_version > self.snapshot.version() {
+            return Err(Error::Generic(format!(
+                "hint_version {} is greater than current version {}",
+                hint_version,
+                self.snapshot.version()
+            )));
+        }
+
+        // in order to be processed by our log replay, we must re-shape the existing scan metadata
+        // back into shape as we read it from the log. Since it is already reconciled data,
+        // we treat it as if it originated from a checkpoint.
+        let transform = engine.evaluation_handler().new_expression_evaluator(
+            Arc::new(scan_row_schema()),
+            get_scan_metadata_transform_expr(),
+            RESTORED_ADD_SCHEMA.clone(),
+        );
+        let apply_transform =
+            move |data: Box<dyn EngineData>| Ok((transform.evaluate(data.as_ref())?, false));
+
+        // If the snapshot version corresponds to the hint version, we process the existing data
+        // to apply file skipping and provide the required transformations.
+        if hint_version == self.snapshot.version() {
+            return Ok(Box::new(self.scan_metadata_inner(
+                engine,
+                hint_data.into_iter().map(apply_transform),
+            )?));
+        }
+
+        // If the current log segment contains a checkpoint newer than the hint version
+        // we disregard the existing data hint, and perform a full scan.
+        if let Some(version) = self.snapshot.log_segment().checkpoint_version {
+            if version > hint_version {
+                let scan_iter = self.scan_metadata(engine)?;
+                return Ok(Box::new(scan_iter));
+            };
+        };
+
+        // create a new log segment containing only the commits added after the version hint.
+        let mut ascending_commit_files = self.snapshot.log_segment().ascending_commit_files.clone();
+        ascending_commit_files.retain(|f| f.version > hint_version);
+        let log_segment = LogSegment::try_new(
+            ascending_commit_files,
+            vec![],
+            self.snapshot.log_segment().log_root.clone(),
+            Some(self.snapshot.log_segment().end_version),
+        )?;
```

Comment:
nit: pull out a local variable:
```rust
let log_segment = self.snapshot.log_segment();`
```
to simplify L522 above as well as this block here:
```suggestion
        let mut ascending_commit_files = log_segment.ascending_commit_files.clone();
        ascending_commit_files.retain(|f| f.version > hint_version);
        let log_segment = LogSegment::try_new(
            ascending_commit_files,
            vec![],
            log_segment.log_root.clone(),
            Some(log_segment.end_version),
        )?;
```